File: src/symbols/backend.py
"""One bounded, non-interactive operation on an explicitly identified module."""

import contextlib
import gdb
import json
import os


def identity(obj):
    # Object addresses can be reused after unload or exec. Tokens cannot.
    token = getattr(obj, "_fgdb_symbol_resolution_identity", None)

    if token is None:
        token = getattr(gdb, "_fgdb_symbol_resolution_sequence", 0) + 1
        gdb._fgdb_symbol_resolution_sequence = token
        obj._fgdb_symbol_resolution_identity = token

    return token


def track_program_instances():
    if hasattr(gdb, "_fgdb_symbol_resolution_clear_handler"):
        return

    def cleared(event):
        space = event.progspace

        if hasattr(space, "_fgdb_symbol_resolution_identity"):
            del space._fgdb_symbol_resolution_identity

    gdb.events.clear_objfiles.connect(cleared)
    gdb._fgdb_symbol_resolution_clear_handler = cleared


def parameter(name, fallback="unavailable"):
    try:
        return gdb.parameter(name)
    except gdb.error:
        return fallback


def emit(values):
    def encode(value):
        if isinstance(value, list):
            return "[" + ",".join(encode(item) for item in value) + "]"
        return json.dumps(str(value), ensure_ascii=False)

    record = "FGDB_SYMBOLS^done," + ",".join(
        name + "=" + encode(value) for name, value in values.items()
    )

    if len(record) > 65536:
        raise gdb.GdbError("Symbol metadata exceeded its output budget")

    gdb.write(record + "\n")


def configuration():
    directories = str(parameter("debug-file-directory", "")).split(os.pathsep)
    directories = list(dict.fromkeys(os.path.abspath(os.path.expanduser(directory))
                                    for directory in extra_directories + directories if directory))

    if len(directories) > 64:
        raise gdb.GdbError("Too many debug-file search directories")

    cache = os.environ.get("DEBUGINFOD_CACHE_PATH")

    if cache is not None:
        caches = [cache]
    else:
        caches = []
        xdg = os.environ.get("XDG_CACHE_HOME")
        home = os.environ.get("HOME")

        if xdg:
            caches.append(os.path.join(xdg, "debuginfod_client"))
        if home:
            caches += [os.path.join(home, ".cache/debuginfod_client"),
                       os.path.join(home, ".debuginfod_client_cache")]

    return {
        "debuginfod": parameter("debuginfod enabled"),
        "urls": parameter("debuginfod urls", os.environ.get("DEBUGINFOD_URLS", "")),
        "auto-solib": int(bool(parameter("auto-solib-add", False))),
        "directories": directories,
        "caches": [os.path.abspath(path) for path in caches if path],
        "cache-override": cache or "",
    }


def library():
    if not hasattr(gdb, "execute_mi"):
        raise gdb.GdbError("Verified module resolution requires GDB's Python MI interface")

    result = gdb.execute_mi("-file-list-shared-libraries", "--thread-group", inferior_id, pattern)
    libraries = result.get("shared-libraries", [])
    matches = [item for item in libraries if item.get("target-name") == target]

    if len(matches) != 1:
        raise gdb.GdbError("The selected module is no longer uniquely mapped")

    item = matches[0]
    ranges = item.get("ranges", [])
    first = ranges[0] if ranges else {}

    for name, expected in (("from", mapping_from), ("to", mapping_to)):
        actual = first.get(name, "")

        if actual != expected:
            raise gdb.GdbError("The selected module mapping changed")

    return item


def object_file(space, item):
    objects = space.objfiles()

    if len(objects) > 20000:
        raise gdb.GdbError("Too many symbol objects to resolve this module safely")

    names = {target, item.get("host-name", "")}
    matches = [obj for obj in objects if obj.is_valid() and obj.owner is None
               and (obj.filename in names or getattr(obj, "username", None) in names)]

    if len(matches) > 1:
        raise gdb.GdbError("More than one symbol object matches the selected module")

    return matches[0] if matches else None


def run():
    values = configuration()

    if action == "configuration":
        emit(values)
        return

    inferior = gdb.selected_inferior()
    space = inferior.progspace
    track_program_instances()

    if "i" + str(inferior.num) != inferior_id:
        raise gdb.GdbError("The selected inferior changed")
    if any(thread.is_running() for thread in inferior.threads()):
        raise gdb.GdbError("Pause all threads before loading debug information")
    if expected_space and (identity(space) != expected_space or inferior.pid != expected_pid):
        raise gdb.GdbError("The module belongs to a different program instance")

    item = library()
    obj = object_file(space, item)

    if expected_object and (obj is None or identity(obj) != expected_object):
        raise gdb.GdbError("The module's symbol object changed")
    if expected_build_id and obj is not None and obj.build_id != expected_build_id:
        raise gdb.GdbError("The loaded module's build ID changed")

    with contextlib.ExitStack() as scope:
        # Network access is owned by fgdb's cancellable download stage.
        if values["debuginfod"] != "unavailable":
            scope.enter_context(gdb.with_parameter("debuginfod enabled", "off"))
        scope.enter_context(gdb.with_parameter("may-call-functions", False))
        scope.enter_context(gdb.with_parameter("debug-file-directory", os.pathsep.join(values["directories"])))

        if action == "load":
            # sharedlibrary consumes a raw regex, not shell-quoted words.
            gdb.execute("sharedlibrary " + pattern, from_tty=False)
        elif action == "attach":
            if obj is None or not hasattr(obj, "add_separate_debug_file"):
                raise gdb.GdbError("This GDB cannot attach a separate debug file to this module")

            stat = os.stat(debug_file)
            actual = ":".join(str(value) for value in
                              (stat.st_dev, stat.st_ino, stat.st_size, stat.st_mtime_ns))

            if actual != debug_stamp:
                raise gdb.GdbError("The debug file changed after validation")

            obj.add_separate_debug_file(debug_file)

    item = library()
    obj = object_file(space, item)
    debug_files = [entry.filename for entry in space.objfiles()
                   if obj is not None and entry.is_valid() and entry.owner == obj]

    if len(debug_files) > 32:
        raise gdb.GdbError("Too many separate debug files for this module")

    values.update({
        "programspace": identity(space), "pid": inferior.pid,
        "object": identity(obj) if obj is not None else 0,
        "loaded": int(str(item.get("symbols-loaded", "0")) == "1"),
        "build-id": (obj.build_id or "") if obj is not None else "",
        "filename": obj.filename if obj is not None else item.get("host-name", target),
        "debug-files": debug_files,
    })
    emit(values)


run()

File: src/symbols/backend.rs
use std::{fmt::Write as _, path::PathBuf};

use super::ModuleKey;
use crate::{
    debug_info::DebugFileSearch,
    debugger::{MiListItem, MiRecord, MiValue, parse_record},
};

#[derive(Clone, Debug, PartialEq, Eq)]
pub(crate) struct Configuration {
    pub debuginfod: String,
    pub urls: String,
    pub auto_solib: bool,
    pub search: DebugFileSearch,
    pub cache_override: Option<PathBuf>,
}

#[derive(Clone, Debug)]
pub(super) struct Snapshot {
    pub configuration: Configuration,
    pub space: u64,
    pub pid: u32,
    pub object: u64,
    pub loaded: bool,
    pub build_id: Option<String>,
    pub filename: PathBuf,
    pub debug_files: Vec<PathBuf>,
}

pub(super) enum Action<'a> {
    Configuration,
    Inspect,
    Load,
    Attach {
        path: &'a std::path::Path,
        stamp: &'a str,
    },
}

pub(super) fn exact_regex(value: &str) -> Result<String, String> {
    if value.is_empty() || value.len() > 16 * 1024 || value.contains(['\0', '\n', '\r']) {
        return Err(String::from("Invalid module path for symbol loading"));
    }

    let mut pattern = String::with_capacity(value.len() + 2);
    pattern.push('^');

    // MI uses a basic regex while CLI commands can use extended syntax.
    // Bracket literals avoid turning a literal '+' into the BRE '\+' operator.
    for character in value.chars() {
        match character {
            '[' => pattern.push_str("[[]"),
            ']' => pattern.push_str("[]]"),
            '^' => pattern.push_str("\\^"),
            '\\' => pattern.push_str("[\\\\]"),
            '.' | '$' | '*' | '+' | '?' | '(' | ')' | '{' | '}' | '|' => {
                pattern.push('[');
                pattern.push(character);
                pattern.push(']');
            }

            _ => pattern.push(character),
        }
    }

    pattern.push('$');
    Ok(pattern)
}

fn python_string(value: &str) -> String {
    let mut literal = String::with_capacity(value.len() * 2 + 28);
    literal.push_str("bytes.fromhex('");

    for byte in value.bytes() {
        let _ = write!(literal, "{byte:02x}");
    }

    literal.push_str("').decode('utf-8')");
    literal
}

pub(super) fn command(
    key: Option<&ModuleKey>,
    action: Action<'_>,
    expected: Option<&Snapshot>,
    directories: &[PathBuf],
) -> Result<String, String> {
    let mut script = String::new();
    let (name, path, stamp) = match action {
        Action::Configuration => ("configuration", "", ""),
        Action::Inspect => ("inspect", "", ""),
        Action::Load => ("load", "", ""),
        Action::Attach { path, stamp } => (
            "attach",
            path.to_str().ok_or("The debug-file path is not UTF-8")?,
            stamp,
        ),
    };

    let _ = writeln!(script, "action = '{name}'");
    let _ = writeln!(
        script,
        "extra_directories = [{}]",
        directories
            .iter()
            .map(|path| path
                .to_str()
                .map(python_string)
                .ok_or("The debug-directory path is not UTF-8"))
            .collect::<Result<Vec<_>, _>>()?
            .join(",")
    );

    if let Some(key) = key {
        let pattern = exact_regex(&key.target)?;

        for (name, value) in [
            ("inferior_id", key.inferior.as_str()),
            ("target", &key.target),
            ("mapping_from", key.from.as_deref().unwrap_or("")),
            ("mapping_to", key.to.as_deref().unwrap_or("")),
            ("pattern", &pattern),
            ("debug_file", path),
            ("debug_stamp", stamp),
            (
                "expected_build_id",
                expected
                    .and_then(|snapshot| snapshot.build_id.as_deref())
                    .unwrap_or(""),
            ),
        ] {
            let _ = writeln!(script, "{name} = {}", python_string(value));
        }

        let _ = writeln!(
            script,
            "expected_space = {}\nexpected_pid = {}\nexpected_object = {}",
            expected.map_or(0, |snapshot| snapshot.space),
            expected.map_or(0, |snapshot| snapshot.pid),
            expected.map_or(0, |snapshot| snapshot.object)
        );
    }

    script.push_str(include_str!("backend.py"));
    Ok(format!("python exec({}, {{}})", python_string(&script)))
}

fn record(output: &str) -> Result<MiRecord, String> {
    if output.len() > 128 * 1024 {
        return Err(String::from("Symbol metadata exceeded its output budget"));
    }

    let mut records = output
        .lines()
        .filter_map(|line| line.strip_prefix("FGDB_SYMBOLS"));

    let record = records
        .next()
        .and_then(|record| parse_record(record).ok())
        .filter(MiRecord::is_done)
        .ok_or("GDB did not return a complete symbol snapshot")?;

    if records.next().is_some() {
        return Err(String::from("GDB returned ambiguous symbol metadata"));
    }

    Ok(record)
}

fn text<'a>(record: &'a MiRecord, key: &str) -> Result<&'a str, String> {
    record
        .field(key)
        .and_then(MiValue::as_const)
        .ok_or_else(|| format!("Missing symbol metadata field '{key}'"))
}

fn flag(record: &MiRecord, key: &str) -> Result<bool, String> {
    match text(record, key)? {
        "0" => Ok(false),
        "1" => Ok(true),
        _ => Err(format!("Invalid symbol metadata flag '{key}'")),
    }
}

fn paths(record: &MiRecord, key: &str, limit: usize) -> Result<Vec<PathBuf>, String> {
    let values = record
        .field(key)
        .and_then(MiValue::as_list)
        .filter(|values| values.len() <= limit)
        .ok_or_else(|| format!("Invalid symbol path list '{key}'"))?;

    values
        .iter()
        .map(|item| match item {
            MiListItem::Value(MiValue::Const(value)) if !value.contains(['\0', '\n', '\r']) => {
                Ok(PathBuf::from(value))
            }

            _ => Err(String::from("Invalid debug-file path")),
        })
        .collect()
}

fn configuration_from(record: &MiRecord) -> Result<Configuration, String> {
    let cache = text(record, "cache-override")?;

    Ok(Configuration {
        debuginfod: text(record, "debuginfod")?.to_owned(),
        urls: text(record, "urls")?.to_owned(),
        auto_solib: flag(record, "auto-solib")?,
        search: DebugFileSearch {
            directories: paths(record, "directories", 64)?,
            caches: paths(record, "caches", 4)?,
        },
        cache_override: (!cache.is_empty()).then(|| PathBuf::from(cache)),
    })
}

pub(super) fn parse_configuration(output: &str) -> Result<Configuration, String> {
    configuration_from(&record(output)?)
}

pub(super) fn parse(output: &str) -> Result<Snapshot, String> {
    let record = record(output)?;
    let number = |key| {
        text(&record, key)?
            .parse::<u64>()
            .map_err(|_| format!("Invalid symbol metadata '{key}'"))
    };

    let build_id = text(&record, "build-id")?;

    if !build_id.is_empty() && !valid_build_id(build_id) {
        return Err(String::from("GDB reported an invalid build ID"));
    }

    Ok(Snapshot {
        configuration: configuration_from(&record)?,
        space: number("programspace")?,
        pid: u32::try_from(number("pid")?).map_err(|_| "Invalid inferior PID")?,
        object: number("object")?,
        loaded: flag(&record, "loaded")?,
        build_id: (!build_id.is_empty()).then(|| build_id.to_owned()),
        filename: PathBuf::from(text(&record, "filename")?),
        debug_files: paths(&record, "debug-files", 32)?,
    })
}

pub(super) fn valid_build_id(value: &str) -> bool {
    (4..=256).contains(&value.len())
        && value.len().is_multiple_of(2)
        && value.bytes().all(|byte| byte.is_ascii_hexdigit())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn exact_module_patterns_are_literal_in_both_gdb_regex_dialects() {
        assert_eq!(
            exact_regex("/lib/a [x]+(v)?{2}|$.so").unwrap(),
            "/lib/a [[]x[]][+][(]v[)][?][{]2[}][|][$][.]so$".replacen('/', "^/", 1)
        );

        assert_eq!(
            exact_regex("/lib/quote\"é.so").unwrap(),
            "^/lib/quote\"é[.]so$"
        );

        assert_eq!(
            exact_regex("/lib/back\\slash^x.so").unwrap(),
            "^/lib/back[\\\\]slash\\^x[.]so$"
        );

        for invalid in ["", "/lib/foo\nsharedlibrary", "/lib/foo\r", "/lib/foo\0"] {
            assert!(exact_regex(invalid).is_err());
        }
    }

    #[test]
    fn only_complete_bounded_symbol_snapshots_are_accepted() {
        let output = "FGDB_SYMBOLS^done,debuginfod=\"ask\",urls=\"\",auto-solib=\"0\",directories=[\"/debug\"],caches=[],cache-override=\"\",programspace=\"123\",pid=\"42\",object=\"321\",loaded=\"1\",build-id=\"1234abcd\",filename=\"/lib/quote\\\"é.so\",debug-files=[\"/debug/a.debug\"]";
        let snapshot = parse(output).unwrap();
        assert!(snapshot.loaded);
        assert_eq!(snapshot.filename, PathBuf::from("/lib/quote\"é.so"));
        assert_eq!(
            snapshot.configuration.search.directories,
            [PathBuf::from("/debug")]
        );

        assert_eq!(snapshot.debug_files, [PathBuf::from("/debug/a.debug")]);
        assert!(parse("^done").is_err());
        assert!(parse(&format!("{output}\n{output}")).is_err());
        assert!(parse(&output.replace("1234abcd", "../invalid")).is_err());
        assert!(parse(&output.replace("loaded=\"1\"", "loaded=\"maybe\"")).is_err());
        assert!(parse(&"x".repeat(128 * 1024 + 1)).is_err());
    }
}
